mitiq/cdr/data_regression.py

```diff
@@ -0,0 +1,294 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+""" Functions for using near-Clifford training data for error mitigation"""
+from typing import List, Union, Callable, Dict
+
+import numpy as np
+
+from cirq.circuits import Circuit
+
+from scipy.optimize import curve_fit
+
+
+def scale_noise_in_circuits(
+    circuits: Union[Circuit, List[Circuit]],
+    scale_noise: Callable[[Circuit, float], Circuit],
+    scale_factor: Union[float, List],
+) -> List[Circuit]:
+    """Returns a list consisting of the original circuit followed by the circuit
+    folded by the specified method and the scale factors.
+
+    Args:
+        circuits: list of training circuits.
+        scale_factor: factor by which to scale the circuits.
+        scale_noise: method to use to fold the circuits.
+    """
+    if isinstance(circuits, Circuit):
+        circuits = [circuits]
+    folded_circuits = []
+    if isinstance(scale_factor, float) or isinstance(scale_factor, int):
+        scale_factor = [scale_factor]
+    for scale_factor_ in scale_factor:
+        for circuit in circuits:
+            folded_circuits.append(scale_noise(circuit, scale_factor_))
+    return circuits + folded_circuits
+
+
+def execute_training_circuits(
+    all_training_circuits: List[Circuit],
+    executor: Callable[[Circuit], Dict],
+    simulator: Callable[[Circuit], Union[Dict, np.ndarray]],
+    noise_levels: int = 1,
+) -> (List[Dict], List[Dict]):
+    """Returns two list of dictionaries of different sizes. One list of
+    dictionaries consisting of the real counts at various noise levels for the
+    training circuits and the other contains the exact counts.
+
+    Args:
+        all_training_circuits: list of all training circuits at all noise
+                               levels.
+        executor: user defined executor function.
+        simulator: user defined simulator function either returns counts or
+                   a statevector.
+        noise_levels: number of noise levels to be used in the regression.
+    """
+    number_of_training_circuits = int(
+        len(all_training_circuits) / noise_levels
+    )
+    training_circuits_raw_data = []
+    training_circuits_simulated_data = []
+    for i, training_circuit in enumerate(all_training_circuits):
+        training_circuit_raw_result = executor(training_circuit)
+        training_circuits_raw_data.append(training_circuit_raw_result)
+        # runs the circuits with no increased noise in the simulator:
+        if i < number_of_training_circuits:
+            training_circuit_simulated_result = simulator(training_circuit)
+            training_circuits_simulated_data.append(
+                training_circuit_simulated_result
+            )
+    return (training_circuits_raw_data, training_circuits_simulated_data)
+
+
+# TODO: this function below is not really necessary, but is quite useful and
+# follows the structure of the above format.
+def execute_circuit_of_interest(
```

Comment:
Again I don't think you really want or need a helper function for this. The way it's called by `execute_with_cdr`, you just need

```python
scaled_circuits = [scale_noise(circuit, s) for s in scale_factors]
results_dict_circuit_of_interest = [executor(c) for c in scaled_circuits]
```